### PR TITLE
feat: undo redo step creation

### PR DESCRIPTION
**Pull request by Kneesal**
# Description

### Issue

creating steps from dragging nodes and New Step button should have undo/redo functionality 

[Link to Basecamp Todo](https://3.basecamp.com/3105655/projects)

### CODE REVIEW

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStep/useCreateStep.ts [hunk 1]
```diff
@@ -1,148 +1,268 @@
...
+import { stepAndCardBlockCreateCacheUpdate } from '../../../../../../libs/useStepAndCardBlockCreateMutation'
 
-type RawEdgeSourceAndCoordinates = RawEdgeSource & {
+type SourceStepAndCoordinates = {
   x: number
   y: number
+  sourceStep: TreeBlock<StepBlock> | null | undefined
 }
 
+export const BLOCK_DELETE_WITH_STEP_UPDATE = gql`
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStep/useCreateStep.ts:42]: STEP_BLOCK_DELETE

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStep/useCreateStep.ts [hunk 2]
```diff
@@ -1,148 +1,268 @@
...
+      }
+    }
+    stepBlockUpdate(id: $stepBlockUpdateId, journeyId: $journeyId, input: $input) {
+      id
+      nextBlockId
+    }
+  }
+`
+
+export const BLOCK_RESTORE_WITH_STEP_UPDATE = gql`
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStep/useCreateStep.ts:58]: STEP_BLOCK_RESTORE

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStep/useCreateStep.ts [hunk 3]
```diff
@@ -1,148 +1,268 @@
...
+      y
+    }
+  }
+  stepBlockUpdate(id: $stepBlockUpdateId, journeyId: $journeyId, input: $input) {
+      id
+      nextBlockId
+    }
+}`
+
+export const STEP_AND_CARD_BLOCK_CREATE_WITH_STEP_UPDATE = gql`
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStep/useCreateStep.ts:76]:
```suggestion
export const STEP_BLOCK_CREATE = gql`
```

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+} from '../../../../../../../__generated__/StepAndCardBlockCreateWithBlockActionUpdate'
+import {
+  ThemeMode,
+  ThemeName
+} from '../../../../../../../__generated__/globalTypes'
+import { blockDeleteUpdate } from '../../../../../../libs/blockDeleteUpdate'
+import { blockRestoreUpdate } from '../../../../../../libs/useBlockRestoreMutation'
+import { stepAndCardBlockCreateCacheUpdate } from '../../../../../../libs/useStepAndCardBlockCreateMutation'
+
+export const BLOCK_DELETE_WITH_BLOCK_ACTION_UPDATE = gql`
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts:37]:
```suggestion
export const STEP_BLOCK_DELETE_FROM_ACTION = gql`
```

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromSocialPreview/useCreateStepFromSocialPreview.ts [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+import { useBlockDeleteMutation } from '../../../../../../libs/useBlockDeleteMutation'
+import { blockRestoreUpdate } from '../../../../../../libs/useBlockRestoreMutation'
+import { stepAndCardBlockCreateCacheUpdate } from '../../../../../../libs/useStepAndCardBlockCreateMutation'
+
+type Coordinates = {
+  x: number
+  y: number
+}
+
+export const STEP_AND_CARD_BLOCK_CREATE_WITH_BLOCK_ORDER_UPDATE = gql`
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromSocialPreview/useCreateStepFromSocialPreview.ts:41]: STEP_BLOCK_CREATE_FROM_SOCIAL_PREVIEW

File: apps/journeys-admin/src/libs/blockDeleteUpdate/blockDeleteUpdate.ts
Comment (tataihono) [inline apps/journeys-admin/src/libs/blockDeleteUpdate/blockDeleteUpdate.ts:1]: revert

**Review by tataihono — CHANGES_REQUESTED** [review]

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStep/useCreateStep.ts [hunk 6]
```diff
@@ -1,148 +1,262 @@
...
+      id
+      nextBlockId
+    }
+  }
+`
 
 export function useCreateStep(): (
-  rawEdgeSource: RawEdgeSourceAndCoordinates
-) => Promise<StepAndCardBlockCreate | null | undefined> {
+  sourceStepAndCoordinates: SourceBlocksAndCoordinates
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStep/useCreateStep.ts:96]:
```suggestion
  input: createStepInput
```

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStep/useCreateStep.ts [hunk 7]
```diff
@@ -1,148 +1,262 @@
...
-                fragment NewBlock on Block {
-                  id
-                }
-              `
-            })
-            return [...existingBlockRefs, newStepBlockRef]
-          }
-        }
-      })
+      stepAndCardBlockCreateCacheUpdate(cache, data, journey?.id)
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStep/useCreateStep.ts:120]: can this go down where the stepBlockCreate is used?

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+    StepBlockRestoreFromAction,
+    StepBlockRestoreFromActionVariables
+  >(STEP_BLOCK_RESTORE_FROM_ACTION)
+
+  return async function createStepFromAction({
+    x,
+    y,
+    sourceStep,
+    sourceBlock
+  }): Promise<void> {
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts:134]: type?

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+        parentBlock {
+            id
+        }
+        gtmEventName
+    }
+  }
+`
+
+export function useCreateStepFromAction(): (
+  sourceBlocksAndCoordinates: SourceBlocksAndCoordinates
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts:105]:
```suggestion
  input: CreateStepInput
```

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+                parentBlockId: sourceBlock.id,
+                __typename: sourceBlock.action.__typename,
+                parentBlock: {
+                  ...sourceBlock
+                },
+                gtmEventName: sourceBlock?.action?.gtmEventName ?? null
+              }
+            },
+            update(cache, { data }, { variables }) {
+              if (data?.blockUpdateAction == null) return
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts:264]: think this should be moved down a line.

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+              parentBlockId: sourceBlock.id,
+              __typename: 'NavigateToBlockAction',
+              gtmEventName: sourceBlock?.action?.gtmEventName ?? null,
+              parentBlock: {
+                ...sourceBlock
+              }
+            }
+          },
+          update(cache, { data }) {
+            if (data?.blockUpdateAction == null) return
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts:310]: this should be moved to 316.

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+                    ...variables?.input
+                  })
+                }
+              })
+            }
+          })
+        }
+      },
+      async redo() {
+        if (sourceBlock == null) return
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts:283]: this check should be done above.

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+      fullscreen: false,
+      coverBlockId: null,
+      backgroundColor: null,
+      parentOrder: 0
+    }
+
+    add({
+      parameters: { execute: {}, undo: { stepBeforeDelete: selectedStep } },
+      async execute() {
+        if (sourceBlock == null) return
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts:163]: this check should be done above

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+          }
+        })
+      },
+      async undo({ stepBeforeDelete }) {
+        if (
+          sourceStep != null &&
+          stepBeforeDelete != null &&
+          sourceBlock != null &&
+          sourceBlock?.action != null
+        ) {
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts:223]: can this check be done above?

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+              parentBlockId: sourceBlock.id,
+              __typename: 'NavigateToBlockAction',
+              gtmEventName: sourceBlock?.action?.gtmEventName ?? null,
+              parentBlock: {
+                ...sourceBlock
+              }
+            }
+          },
+          update(cache, { data }) {
+            if (data?.blockUpdateAction == null) return
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts:203]: move down a line

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStep/useCreateStep.ts [hunk 16]
```diff
@@ -1,148 +1,262 @@
...
-      selectedStep: {
-        ...data.stepBlockCreate,
-        children: [{ ...data.cardBlockCreate, children: [] }]
-      }
-    })
-
-    const edgeSource = convertToEdgeSource(rawEdgeSource)
+      },
+      async undo({ stepBeforeDelete }) {
+        if (sourceStep != null && stepBeforeDelete != null) {
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStep/useCreateStep.ts:197]: this check can be done above.

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStep/useCreateStep.ts [hunk 17]
```diff
@@ -1,148 +1,262 @@
...
+      parameters: {
+        execute: {},
+        undo: { stepBeforeDelete: selectedStep }
+      },
+      async execute() {
+        dispatch({
+          type: 'SetSelectedStepByIdAction',
+          selectedStepId: step.id
+        })
+        if (sourceStep != null) {
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStep/useCreateStep.ts:162]: same for this

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStep/useCreateStep.ts [hunk 18]
```diff
@@ -1,148 +1,262 @@
...
-          optimisticResponse: {
-            stepBlockUpdate: {
-              id: edgeSource.stepId,
-              __typename: 'StepBlock',
-              nextBlockId: newStepId
+          })
+        }
+      },
+      async redo() {
+        if (sourceStep != null) {
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStep/useCreateStep.ts:227]: move up

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStep/useCreateStep.ts (hunk 7 above)
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStep/useCreateStep.ts:120]: can we change the name of this update to stepBlockCreateUpdate

**Review by tataihono — CHANGES_REQUESTED** [review]

**Review by tataihono — CHANGES_REQUESTED** [review]

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts [hunk 19]
```diff
@@ -0,0 +1,330 @@
...
+              }),
+              fields: {
+                action: () => ({ ...data.blockUpdateAction, blockId: step.id })
+              }
+            })
+          }
+        })
+      },
+      async undo({ stepBeforeDelete }) {
+        if (sourceBlock.action == null) return
```
[the remark this answers is not in the document]
Comment (Kneesal) [reply apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/libs/useCreateStepFromAction/useCreateStepFromAction.ts:225]: can't move out

**Review by tataihono — APPROVED** [review]

**Pull request merged**